core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -285,6 +287,18 @@ public Mapper parse(ParseContext context) throws IOException {
      */
     protected abstract void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException;
 
+    protected void createFieldNamesField(ParseContext context, List<IndexableField> fields) {
+        FieldNamesFieldType fieldNamesFieldType = (FieldNamesFieldMapper.FieldNamesFieldType) context.docMapper()
+                .metadataMapper(FieldNamesFieldMapper.class).fieldType();
+        if (fieldNamesFieldType != null && fieldNamesFieldType.isEnabled()) {
+            for (String fieldName : FieldNamesFieldMapper.extractFieldNames(fieldType().name())) {
+                if (fieldType().indexOptions() != IndexOptions.NONE || fieldType().stored()) {
```

Comment:
@jpountz and I discussed this and we agreed that for this PR we will keep the behaviour that exists queries will only work for fields that are indexed, stored, or have doc values (or a combination of those) so this PR can be merged into 7.0 and 6.x. I'll make a followup PR for 7.0 only that changes things to allow exists to work for any field that is mapped.

Reply:
Hmm but then that if statement is still buggy since it should also check for `hasDocValues`? Actually I'd be in favour of removing it entirely since you seem to have that logic on the call sites already anyway (only adding this field if there are no doc values)?

Reply:
I've moved this logic outside of this method, the callers now check to ensure if field names field should be created